cmake/extensions.cmake

```diff
@@ -767,6 +774,148 @@ function(board_finalize_runner_args runner)
   set_property(GLOBAL APPEND PROPERTY ZEPHYR_RUNNERS ${runner})
 endfunction()
 
+# Zephyr board revision:
+#
+# This section provides a function for revision checking.
+
+# Usage:
+#   board_check_revision(FORMAT <LETTER | MAJOR.MINOR.PATCH>
+#                        [EXACT]
+#                        [DEFAULT_REVISION <revision>]
+#                        [HIGHEST_REVISION <revision>]
+#   )
+#
+# Zephyr board extension function.
+#
+# This function can be used in `boards/<board>/revision.cmake` to check a user
+# requested revision against available board revisions.
+#
+# The function will check the revision from `-DBOARD=<board>@<revision>` that
+# is provided by the user according to the arguments.
+# When `EXACT` is not specified, this function with set the Zephyr build system
+# variable `ACTIVE_BOARD_REVISION` with the selected revision.
+#
+# FORMAT <LETTER | MAJOR.MINOR.PATCH>: Specify the revision format.
+#         LETTER:             Revision format is a single letter from A - Z.
+#         MAJOR.MINOR.PATCH:  Revision format is three digits, separated by `.`,
+#                             `x.y.z`. Trailing zeroes may be omitted on the
+#                             command line, which means:
+#                             1.0.0 == 1.0 == 1
+#
+# EXACT: Revision is required to be an exact match. As example, available revisions are:
+#        0.1.0 and 0.3.0, and user provides 0.2.0, then an error is reported
+#        when `EXACT` is given.
+#        If `EXACT` is not provided, then closest lower revision will be selected
+#        as the active revision, which in the example will be `0.1.0`.
+#
+# DEFAULT_REVISION: Provides a default revision to use when user has not selected
+#                   a revision number. If no default revision is provided then
+#                   user will be printed with an error if no revision is given
+#                   on the command line.
+#
+# HIGHEST_REVISION: Allows to specify highest valid revision for a board.
+#                   This can be used to ensure that a newer board cannot be used
+#                   with an older Zephyr. As example, if current board supports
+#                   revisions 0.x.0-0.99.99 and 1.0.0-1.99.99, and it is expected
+#                   that current board implementation will not work with board
+#                   revision 2.0.0, then HIGHEST_REVISION can be set to 1.99.99,
+#                   and user will be printed with an error if using
+#                   `<board>@2.0.0` or higher.
+#                   This field is not needed when `EXACT` is used.
+#
+function(board_check_revision)
```

Comment:
That being said, it would be good to have a function `board_revision_check` which you could call from a samples cmakelist (or app specific revision.cmake) to ensure that the current board revision satisfies some criteria (much like this function already does, but only checking the version itself, not the existance of configuration files and setting board variables etc)

Reply:
no, the function is checking the revision provided by the user and will print an error if that revision is not valid.

It doesn't change the value of the revision, 1 == 1.0 == 1.0.0.
It sets an internal build variable `ACTIVE_BOARD_REVISION`, but that is build internals and is not changing revision, which is done later.

> board revision satisfies some criteria

In order to implement such a check, there would be a need for knowing some more specific criteria.
With `BOARD_REVISION` in place, a sample can start doing:
```
if(${BOARD} STREQUAL plank AND ${BOARD_REVISION} VERSION_LESS 1.2.3)
  do_something()
endif()
```
isn't that sufficient ?

Reply:
I was not aware of the built-in support for `VERSION_LESS`, that should suffice.